src/mono/mono/utils/mono-wasm-pagemgr.c

```diff
@@ -0,0 +1,409 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include <config.h>
+#include "mono-wasm-pagemgr.h"
+#include <memory.h>
+#include <unistd.h>
+#include <glib.h>
+#include <mono/utils/atomic.h>
+#include <mono/utils/mono-os-mutex.h>
+
+// #define MWPM_LOGGING
+// #define MWPM_STATS
+
+typedef enum {
+	MWPM_UNINITIALIZED = 0,
+	MWPM_INITIALIZING = 1,
+	MWPM_INITIALIZED = 2
+} init_state;
+
+typedef enum {
+	MWPM_MARK_DEAD_PAGES,
+	MWPM_MARK_NEW_PAGES,
+	MWPM_FREE_TO_ALLOCATED,
+	MWPM_FREE_TO_ALLOCATED_ZEROED,
+	MWPM_ALLOCATED_TO_FREE,
+} page_action;
+
+#define is_page_free(state) (state & MWPM_FREE_BIT)
+#define is_page_owned(state) (state & MWPM_STATE_MASK)
+#define is_page_in_use(state) ((state & MWPM_STATE_MASK) == MWPM_ALLOCATED)
+#define get_page_skip_count(state) (state & MWPM_SKIP_MASK) + 1
+
+typedef uint8_t mwpm_page_state;
+
+static mono_mutex_t mutex;
+static uint8_t page_table[MWPM_MAX_PAGES];
+static gint32 is_initialized = MWPM_UNINITIALIZED;
+static uint32_t
+	// The index of the first page that we control. Not all pages after this
+	//  necessarily belong to us, but scans can start here.
+	first_controlled_page_index = UINT32_MAX,
+	// The index of the last page we've allocated. Not all pages between this
+	//  and first_controlled_page_index belong to us, but scans can end here.
+	last_controlled_page_index = 0;
+static uint8_t *prev_waste_start = NULL,
+	*prev_waste_end = NULL;
+
+static inline void *
+address_from_page_index (uint32_t page_index) {
+	uint64_t address = ((uint64_t)page_index * MWPM_PAGE_SIZE);
+	g_assert (address < UINT32_MAX);
+	return (void *)(uint32_t)address;
+}
+
+static inline uint32_t
+first_page_from_address (void *addr) {
+	return ((uint64_t)addr) / MWPM_PAGE_SIZE;
+}
+
+static inline uint32_t
+page_count_from_size (size_t size) {
+	return ((size + MWPM_PAGE_SIZE - 1) / MWPM_PAGE_SIZE);
+}
+
+static inline uint32_t
+last_page_of_range (void *addr, size_t size) {
+	uint32_t page_count_rounded_up = page_count_from_size (size),
+		first_page = first_page_from_address (addr);
+	return first_page + page_count_rounded_up - 1;
+}
+
+static inline mwpm_page_state
+encode_page_state (uint8_t bits, uint32_t skip_count) {
+	if (skip_count > MWPM_SKIP_MASK)
+		skip_count = MWPM_SKIP_MASK;
```

Comment:
I think this is a bit subtle and might be worth a comment.

I think what happens is that if we have a run of pages that are all going to be skipped, `transition_page_states` will mark all of them with something like:

```
[DIRTY|MAX_SKIP]...[DIRTY|MAX_SKIP][DIRTY|MAX_SKIP-1][DIRTY|MAX_SKIP-2]...[DIRTY|2][DIRTY|1][DIRTY|0]
```

so if you have let's say `MAX_SKIP+2` dirty pages, if you start looking starting from the first or second or third of those dirty pages, it will make you skip ahead `MAX_SKIP` entries but you'll still be in the same range, and the later pages (once the skip count is below the maximum representable skip) will help you get aligned correctly.

Reply:
Yeah, I should add a detailed comment somewhere explaining the skip algorithm. Good point.